File: algoritms/breadthsearch.php
<?php 

$graph = array(
    'you' => array('alice', 'bob', 'claire'),
    'bob' => array('anuj', 'peggy'),
    'alice' => array('peggy'),
    'claire' => array('thom', 'jonny'),
    'anuj' => array(),
    'peggy' => array(),
    'thom' => array(),
    'jonny' => array(),  
);

function person_is_seller($name) {
    $value = substr($name, -1);
    return strcasecmp($value,'m'); 
}
    
function searchseller($name, $graph) {
      
    $searchQueue = array();
    $searched = array();
        
    $searchQueue[$name] = $graph[$name];
    
    $n = 0;
    $c = count($searchQueue);
       
    while ($n < $c) {
        $keys = array_keys($searchQueue);
        $key = $keys[$n];
        $value = $searchQueue[$key];
        $n += 1;
        
        if(!isset($searched[$key])) {  
            if(!person_is_seller($key)) {
                echo "$key is a mango seller";
                break;
            } else {
                unset($searchQueue[$key]);
                foreach ($value as $k => $v) {
                    $searchQueue[$v] = $graph[$v];
                }
                $searched[$key] = array();
                $c = count($searchQueue);
                $n = 0;       
            }
            
        } else {
            echo "end";
        }
        
    }
        
}
    
searchseller('you', $graph);

File: algoritms/BreadthSearch.java
package algoritms;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Queue;

public class BreadthSearch {
	
	private static Map<String, List<String>> graph = new HashMap<>();
	
	private static boolean search(String name) {
		
		Queue<String> searchQueue = new ArrayDeque<>(graph.get(name));
		List<String> searched = new ArrayList<>();
		
		while(!searchQueue.isEmpty()) {
			String person = searchQueue.poll();
			if(!searched.contains(person)) {
				if(person_is_seller(person)) {
					System.out.println(person + " is a mango seller");
				} else {
					searchQueue.addAll(graph.get(person));
					searched.add(person);
				}
			}
		}
		
		return false;
		
	}
	
	private static boolean person_is_seller(String name) {
		return name.endsWith("m");
	}

	public static void main(String[] args) {
		
		graph.put("you", Arrays.asList("alice", "bob", "claire"));
		graph.put("bob", Arrays.asList("anuj", "peggy"));
		graph.put("alice", Arrays.asList("peggy"));
		graph.put("claire", Arrays.asList("thom", "jonny"));
		graph.put("anuj", Collections.emptyList());
		graph.put("peggy", Collections.emptyList());
		graph.put("thom", Collections.emptyList());
		graph.put("jonny", Collections.emptyList());
		
		search("you");
		
	}

}
